stylo: Store location information for keyframe rules.

chenpighead opened the pull request:
So far, we only store location info for the whole Keyframes block, not for each
of the keyframe rule. Without this info, the devtool can't present the rules
on the devtool panel properly.

In this patch, we collect the source location info while parsing keyframe selector.
The binding function, Servo_KeyframesRule_GetKeyframe, is also fixed (and renamed
to Servo_KeyframesRule_GetKeyframeAt to match the fix) to accept line and column
parameters from Gecko, so we can pass/set them with the source location from Servo.

This is the servo part of [Bug 1394994](https://bugzilla.mozilla.org/show_bug.cgi?id=1394994).

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix [Bug 1394994](https://bugzilla.mozilla.org/show_bug.cgi?id=1394994)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/18365)
<!-- Reviewable:end -->

chenpighead committed:
stylo: Store location information for keyframe rules.

So far, we only store location info for the whole Keyframes block, not for each
of the keyframe rule. Without this info, the devtool can't present the rules
on the devtool panel properly.

In this patch, we collect the source location info while parsing keyframe selector.
The binding function, Servo_KeyframesRule_GetKeyframe, is also fixed (and renamed
to Servo_KeyframesRule_GetKeyframeAt to match the fix) to accept line and column
parameters from Gecko, so we can pass/set them with the source location from Servo.

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/stylesheets/keyframes_rule.rs, components/style/stylesheets/rule_parser.rs
 * @canaltinova: components/style/stylesheets/keyframes_rule.rs, components/style/stylesheets/rule_parser.rs
 * @emilio: ports/geckolib/glue.rs, components/style/stylesheets/keyframes_rule.rs, components/style/stylesheets/rule_parser.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!
* These commits modify style code, but no tests are modified. Please consider adding a test!

chenpighead commented:
This has been reviewed by xidorn in [Bug 1394994.](https://bugzilla.mozilla.org/show_bug.cgi?id=1394994)

Gecko try: https://treeherder.mozilla.org/#/jobs?repo=try&revision=9a189267f4d06a8afffa04a81a79f915d3b05c21

chenpighead commented:
@bors-servo r=xidorn

bors-servo commented:
:pushpin: Commit 804c17f has been approved by `xidorn`

<!-- @bors-servo r=xidorn 804c17f3ff9b274d6b380fe570a5f8dba2411ff8 -->

bors-servo commented:
:hourglass: Testing commit 804c17f3ff9b274d6b380fe570a5f8dba2411ff8 with merge 1f520b3d4b791ae80e65d46a1193c2d1ba806851...

bors-servo commented:
:broken_heart: Test failed - [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/4325)

chenpighead commented:
Oops, lack of bindings.rs part I guess, I'll update the patch

chenpighead commented:
@bors-servo r=xidorn

bors-servo commented:
:pushpin: Commit 16a7f48 has been approved by `xidorn`

<!-- @bors-servo r=xidorn 16a7f4802bbce854f3f8e7fcd9018d9f4e983fbc -->

bors-servo commented:
:hourglass: Testing commit 16a7f4802bbce854f3f8e7fcd9018d9f4e983fbc with merge 12b62f0cfeb2b77ac57fcc81539c4a6dc7a6c6e7...

bors-servo commented:
:broken_heart: Test failed - [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/4328)

chenpighead commented:
Sorry, it's the unit test that I forgot to update... :(

chenpighead committed:
stylo: Fix unit test failures caused by adding source_location member to Keyframe rule.

In #18365, we add an extra source_location member to Keyframe rule, so that we
can provide location information to devtool panel for Stylo. In order not to
break the existing unit tests, we should add this new member to all the Keyframe
rules used in our unit tests.

As for the tests in stylesheets.rs, since we make the test by mocking a raw css
string, there are actual source locations that we can add to the expected result.
As to the tests in keyframes.rs, since we don't use raw css string, we can just
add a dummy souce location for all the tests. We can further fix this if we
decide to use a raw css string for the tests in keyframes.rs one day.

emilio commented on tests/unit/style/stylesheets.rs:
```diff
@@ -217,7 +217,11 @@ fn test_parse_stylesheet() {
-                            ])))
+                            ]))),
+                            source_location: SourceLocation {
+                                                 line: 17,
```
nit: Align this four spaces to the right of the "s" in `source_location`

emilio commented on tests/unit/style/stylesheets.rs:
```diff
@@ -231,6 +235,10 @@ fn test_parse_stylesheet() {
                                  Importance::Normal),
                             ]))),
+                            source_location: SourceLocation {
+                                                 line: 18,
```
nit: And here too.

chenpighead replied:
No problem. Thanks for the quick review. :)

emilio reviewed (approved):
Unit test changes lgtm.

chenpighead commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit f9b7bebc442245a73a519375812d12d6f9787faa with merge 7a67936470dba9a7e915ae3541f8de4edaa4d6ac...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/8713), [arm32](http://build.servo.org/builders/arm32/builds/8946), [arm64](http://build.servo.org/builders/arm64/builds/8977), [linux-dev](http://build.servo.org/builders/linux-dev/builds/9037), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/5828), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/5870), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/9073), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/1536), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/1544), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/5771), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/5830), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/1559), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/1553), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/4332)
State: approved= try=True

chenpighead commented:
@bors-servo r=upsuper, emilio

bors-servo commented:
:pushpin: Commit f9b7beb has been approved by `upsuper,`

<!-- @bors-servo r=upsuper, f9b7bebc442245a73a519375812d12d6f9787faa -->

bors-servo commented:
:hourglass: Testing commit f9b7bebc442245a73a519375812d12d6f9787faa with merge 1d0d139c8e53643ee8df03689f1d4eb52c08ef2e...

bors-servo commented:
:broken_heart: Test failed - [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/9081)

chenpighead commented:
wat?
@bors-servo retry

bors-servo commented:
:zap: Previous build results for [android](http://build.servo.org/builders/android/builds/8721), [linux-dev](http://build.servo.org/builders/linux-dev/builds/9045), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/4340) are reusable. Rebuilding only [arm32](http://build.servo.org/builders/arm32/builds/8954), [arm64](http://build.servo.org/builders/arm64/builds/8985), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/5836), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/5878), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/9081), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/1544), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/1551), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/5779), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/5838), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/1567), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/1561)...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/5878)

chenpighead commented:
seriously?
@bors-servo retry

bors-servo commented:
:zap: Previous build results for [android](http://build.servo.org/builders/android/builds/8721), [arm64](http://build.servo.org/builders/arm64/builds/8985), [linux-dev](http://build.servo.org/builders/linux-dev/builds/9045), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/5836), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/1544), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/5779), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/4340) are reusable. Rebuilding only [arm32](http://build.servo.org/builders/arm32/builds/8954), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/5878), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/9082), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/1551), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/5838), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/1567), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/1561)...

bors-servo commented:
:broken_heart: Test failed - [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/9082)

chenpighead commented:
@bors-servo retry

bors-servo commented:
:zap: Previous build results for [android](http://build.servo.org/builders/android/builds/8721), [arm32](http://build.servo.org/builders/arm32/builds/8954), [arm64](http://build.servo.org/builders/arm64/builds/8985), [linux-dev](http://build.servo.org/builders/linux-dev/builds/9045), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/5836), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/1544), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/5779), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/5838), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/1567), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/1561), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/4340) are reusable. Rebuilding only [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/5879), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/9082), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/1551)...

bors-servo commented:
:broken_heart: Test failed - [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/9083)

chenpighead commented:
@bors-servo retry

bors-servo commented:
:zap: Previous build results for [android](http://build.servo.org/builders/android/builds/8721), [arm32](http://build.servo.org/builders/arm32/builds/8954), [arm64](http://build.servo.org/builders/arm64/builds/8985), [linux-dev](http://build.servo.org/builders/linux-dev/builds/9045), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/5836), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/1544), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/1551), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/5779), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/5838), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/1567), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/1561), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/4340) are reusable. Rebuilding only [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/5879), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/9083)...

upsuper commented:
You would fail again with mac-dev-unit. The slave machine building this is somehow broken.

chenpighead commented:
Hmm, I saw a PR that retry couple times and then get passed yesterday, thought I could get some lucky as well... maybe I'm running into a different issue...  :disappointed:

bors-servo commented:
:broken_heart: Test failed - [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/9084)

upsuper commented:
@bors-servo retry

bors-servo commented:
:zap: Previous build results for [android](http://build.servo.org/builders/android/builds/8721), [arm32](http://build.servo.org/builders/arm32/builds/8954), [arm64](http://build.servo.org/builders/arm64/builds/8985), [linux-dev](http://build.servo.org/builders/linux-dev/builds/9045), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/5836), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/5879), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/1544), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/1551), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/5779), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/5838), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/1567), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/1561), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/4340) are reusable. Rebuilding only [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/9084)...

upsuper commented:
That machine seems to really like your PR.

bors-servo commented:
:broken_heart: Test failed - [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/9085)

chenpighead commented:
@bors-servo retry

bors-servo commented:
:zap: Previous build results for [android](http://build.servo.org/builders/android/builds/8721), [arm32](http://build.servo.org/builders/arm32/builds/8954), [arm64](http://build.servo.org/builders/arm64/builds/8985), [linux-dev](http://build.servo.org/builders/linux-dev/builds/9045), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/5836), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/5879), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/1544), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/1551), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/5779), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/5838), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/1567), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/1561), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/4340) are reusable. Rebuilding only [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/9085)...

bors-servo commented:
:broken_heart: Test failed - [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/9086)

chenpighead commented:
@bors-servo r-

chenpighead commented:
@bors-servo r=upsuper,emilio

bors-servo commented:
:pushpin: Commit f9b7beb has been approved by `upsuper,emilio`

<!-- @bors-servo r=upsuper,emilio f9b7bebc442245a73a519375812d12d6f9787faa -->

chenpighead commented:
Filed #18374 for the failure...

chenpighead commented:
@bors-servo retry p=1

chenpighead commented:
@bors-servo retry

chenpighead commented:
@bors-servo r=upsuper,emilio

bors-servo commented:
:pushpin: Commit f9b7beb has been approved by `upsuper,emilio`

<!-- @bors-servo r=upsuper,emilio f9b7bebc442245a73a519375812d12d6f9787faa -->

bors-servo commented:
:bulb: This pull request was already approved, no need to approve it again.

- There's another pull request that is currently being tested, blocking this pull request: #18271

bors-servo commented:
:pushpin: Commit f9b7beb has been approved by `upsuper,emilio`

<!-- @bors-servo r=upsuper,emilio f9b7bebc442245a73a519375812d12d6f9787faa -->

bors-servo commented:
:hourglass: Testing commit f9b7bebc442245a73a519375812d12d6f9787faa with merge f648e12935cb94405a815216391f2527912ba4c2...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/8729), [arm32](http://build.servo.org/builders/arm32/builds/8962), [arm64](http://build.servo.org/builders/arm64/builds/8993), [linux-dev](http://build.servo.org/builders/linux-dev/builds/9053), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/5844), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/5888), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/9094), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/1552), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/1559), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/5789), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/5846), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/1577), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/1570), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/4348)
Approved by: upsuper,emilio
Pushing f648e12935cb94405a815216391f2527912ba4c2 to master...
